Zend/zend_string.c

```diff
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
-
-	zend_new_interned_string = zend_new_interned_string_int;
-	zend_interned_strings_snapshot = zend_interned_strings_snapshot_int;
-	zend_interned_strings_restore = zend_interned_strings_restore_int;
+	zend_known_strings = pemalloc(sizeof(zend_string*) * ((sizeof(known_strings) / sizeof(known_strings[0]) - 1)), 1);
+	for (i = 0; i < (sizeof(known_strings) / sizeof(known_strings[0])) - 1; i++) {
+		str = zend_string_init(known_strings[i], strlen(known_strings[i]), 1);
+		zend_known_strings[i] = zend_new_interned_string_permanent(str);
+	}
 }
 
-void zend_interned_strings_dtor(void)
+ZEND_API void zend_interned_strings_dtor(void)
 {
-#ifndef ZTS
-	zend_hash_destroy(&CG(interned_strings));
-#else
-	uint32_t i;
+	zend_hash_destroy(&interned_strings_permanent);
 
-	for (i = 0; i < CG(known_strings_count); i++) {
-		zend_zts_interned_string_free(&CG(known_strings)[i]);
-	}
-#endif
-	free(CG(known_strings));
-	CG(known_strings) = NULL;
-	CG(known_strings_count) = 0;
-	known_interned_strings = NULL;
-	known_interned_strings_count = 0;
+	free(zend_known_strings);
+	zend_known_strings = NULL;
 }
 
-static zend_string *zend_new_interned_string_int(zend_string *str)
+static zend_always_inline zend_string *zend_interned_string_ht_lookup(zend_string *str, HashTable *interned_strings)
 {
-#ifndef ZTS
 	zend_ulong h;
 	uint32_t nIndex;
 	uint32_t idx;
 	Bucket *p;
 
-	if (ZSTR_IS_INTERNED(str)) {
-		return str;
-	}
-
 	h = zend_string_hash_val(str);
-	nIndex = h | CG(interned_strings).nTableMask;
-	idx = HT_HASH(&CG(interned_strings), nIndex);
+	nIndex = h | interned_strings->nTableMask;
+	idx = HT_HASH(interned_strings, nIndex);
 	while (idx != HT_INVALID_IDX) {
-		p = HT_HASH_TO_BUCKET(&CG(interned_strings), idx);
+		p = HT_HASH_TO_BUCKET(interned_strings, idx);
 		if ((p->h == h) && (ZSTR_LEN(p->key) == ZSTR_LEN(str))) {
 			if (!memcmp(ZSTR_VAL(p->key), ZSTR_VAL(str), ZSTR_LEN(str))) {
-				zend_string_release(str);
 				return p->key;
 			}
 		}
 		idx = Z_NEXT(p->val);
 	}
 
-	GC_REFCOUNT(str) = 1;
-	GC_FLAGS(str) |= IS_STR_INTERNED;
-
-	if (CG(interned_strings).nNumUsed >= CG(interned_strings).nTableSize) {
-		if (CG(interned_strings).nTableSize < HT_MAX_SIZE) {	/* Let's double the table size */
-			void *new_data;
-			void *old_data = HT_GET_DATA_ADDR(&CG(interned_strings));
-			Bucket *old_buckets = CG(interned_strings).arData;
-
-			CG(interned_strings).nTableSize += CG(interned_strings).nTableSize;
-			CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-			new_data = malloc(HT_SIZE(&CG(interned_strings)));
-
-			if (new_data) {
-				HT_SET_DATA_ADDR(&CG(interned_strings), new_data);
-				memcpy(CG(interned_strings).arData, old_buckets, sizeof(Bucket) * CG(interned_strings).nNumUsed);
-				free(old_data);
-				zend_hash_rehash(&CG(interned_strings));
-			} else {
-				CG(interned_strings).nTableSize = CG(interned_strings).nTableSize >> 1;
-				CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-			}
+	return NULL;
```

Comment:
As the interned string is stored as both key and value of the HT (https://github.com/weltling/php-src/blob/97882476c602a20411c33d0d5c555818ba5e5834/Zend/zend_string.c#L176) this function should be a thin wrapper around `zend_hash_find()`, instead of reimplementing HT logic.